extensions/resume.md

```diff
@@ -0,0 +1,306 @@
+
+    C: CAP LS
+    S: CAP * LS :multi-prefix draft/resume-0.4
+    C: CAP REQ :draft/resume-0.4
+    S: CAP * ACK :draft/resume-0.4
+    S: RESUME TOKEN JKbAypzFiovffzuD8VEfcs6bOLrXsSenxsyZNt87Bb0YaQlyOzuQciz6cB2R
+
+Considerations around tokens and the process for generating them is described below in the [Resume Token](#resume-token) section.
+
+
+## Resuming A Connection
+When a client detects that it has become disconnected from a server, it SHOULD try to resume before it breaks the existing connection.
+
+Upon establishing the new connection, the client begins capability negotiation, negotiates all mutually-supported capabilities, and MUST confirm that the `draft/resume-0.4` capability exists. If this capability does not exist, the client continues connection registration without attempting to resume. If this capability does exist, the client sends the `RESUME` command and MUST wait for either a `RESUME` or a `FAIL RESUME` message from the server before continuing registration. It should be noted that the client MUST NOT perform SASL authentication if the `draft/resume-0.4` capability exists and they wish to resume their session, as completing SASL auth will end connection registration and abort the resumption attempt.
+
+If the token provided by the new client is validated by the server, the old client completed connection registration with the server, and both the old and new clients use TLS, then the attempt SHOULD be successful. If the attempt is successful, the server MUST send the client a `RESUME` message and complete connection registration immediately (at which time the state will begin to replay as described below). If the attempt is unsuccessful (for example, if either session is not using a secure connection), the server MUST send a `FAIL RESUME` message, and then allow the client to continue connection registration.
+
+If the client receives a `RESUME SUCCESS` message, connection registration will complete immediately and state will begin to replay as described below. If the client receives a `FAIL RESUME`, the client MUST continue registration as though they are joining the server normally.
+
+On a successful request, the server:
+
+1. Terminates the old client's connection.
+2. Updates the client's username, hostname, and resume token (or lack of one) to match the new client's information.
+3. Plays the 'joining server' burst to the new client.
+4. Replays the client's session to the new client.
+5. Sends `RESUMED` or `QUIT`+`JOIN` messages to other clients as appropriate.
+6. Sends clients that have the reconnecting user `MONITOR`'d one `RPL_MONOFFLINE` numeric and one `RPL_MONONLINE` numeric indicating that the user has reconnected. The timestamps on both these messages, if sent, SHOULD be the current time.
+7. If message history could not be replayed (because it is not stored, or for any other reason), sends the new client a `RESUME WARN` message making them aware of this.
```

Comment:
`WARN RESUME`?

Reply:
Also, should we standardize a `WARN RESUME` code for this? I propose `WARN RESUME HISTORY_LOST`.